### PR TITLE
Aspect writers ranges tags

### PR DESCRIPTION
**Pull request by jgraff2**
Maintain keys in redis for writers, ranges, and tags by aspect, to avoid needing to store the entire aspect object. Use the ranges keys to do the status calculation.

### CODE REVIEW

**Comment by jgraff2** [comment]
@pallavi2209 ready for you to use.

**Comment by pallavi2209** [comment]
@jgraff2 The precedence logic looks a bit confusing to me. Lets discuss it whenever possible.

File: tests/db/model/sample/statusCalculation.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ const Aspect = tu.db.Aspect;
 const Subject = tu.db.Subject;
 const Op = require('sequelize').Op;
 
-describe('tests/db/model/sample/statusCalculation.js >', () => {
+describe.skip('tests/db/model/sample/statusCalculation.js >', () => {
```
[the remark this answers is not in the document]
Comment (iamigo) [reply tests/db/model/sample/statusCalculation.js]: maybe create another story for "cleanup" to remove that code and tests once this is all rolled out?

**Comment by pallavi2209** [comment]
@jgraff2 Updated the PR as discussed.

**Comment by jgraff2** [comment]
@pallavi2209 @iamigo 
I've updated the PR with a new approach to status calculation that fully supports the existing behavior by pre-processing the ranges.
This is a significant change to the PR, so please re-review.
Let me know if you want to discuss.

File: cache/statusCalculation.js [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    best = nextTry;
+    adjustment /= 10;
+    nextTry = n - adjustment;
+  }
+
+  return best;
+}
+
+/**
+ * Adjust the given value down by the smallest possible amount that is
```
Comment (iamigo) [inline cache/statusCalculation.js]: up

File: tests/cache/models/samples/statusCalculation.js [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+  } // setupRanges
+
+  function calculateAndExpect(value, expectedStatus) {
+    return redisOps.calculateSampleStatus('sub1.sub2|asp1', value)
+           .should.eventually.equal(expectedStatus);
+  }
+
+  describe('numeric >', () => {
+    describe('basic ranges (AND calculate invoked on update) >', () => {
+      beforeEach(() =>
```
Comment (iamigo) [inline tests/cache/models/samples/statusCalculation.js]: this doesn't need to get set up before *each* test in this "describe", just once, right?

File: tests/cache/models/samples/statusCalculation.js [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+        calculateAndExpect('11', constants.statuses.Invalid)
+      );
+
+      it('value is less than any of the range min values', () =>
+        calculateAndExpect('-1', constants.statuses.Invalid)
+      );
+    });
+
+    describe('negative int ranges with null range in the middle >', () => {
+      beforeEach(() =>
```
Comment (iamigo) [inline tests/cache/models/samples/statusCalculation.js]: this doesn't need to get set up before *each* test in this "describe", just once, right?

**Pull request merged**